juju/juju
# raft-leases: move change notification from the store to the forwarder

## Description of change

When lease changes (claims or expiries) happen in raft, we need to update the database accordingly. This was being done in the raftlease store, but because commands are sent across the hub, it's hard to be certain whether a command has been successfully applied (and thus that the db needs to be updated) - in particular one can't tell the difference between a message being dropped (because of a raft-leadership change) or one where we just didn't get a response in time.

Instead, move the notification to happen in the forwarder, which is on the other side of the hub applying commands to the raft node directly. FSM.Apply now returns a `FSMResponse` which indicates whether the operation was successful, and can update a notification target with any changes.

I initially thought about putting the notification inside the FSM, but that doesn't work - the FSM can't know whether the commands it sees are current commands or replaying log entries to bring a newly-joined raft node up to the current state. So the fact that the forwarder calls `response.Notify(target)` is key to ensuring that only real commands generate DB changes.

## QA steps

* Bootstrap a controller, deploy an application with multiple units, ensure that leadership changes are reflected in the leaseholders collection in mongo (and in the status output).
